File: examples/face_landmarker/ios/FaceLandmarker/Helper/OverlayView.swift
// Copyright 2023 The MediaPipe Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import UIKit
import MediaPipeTasksVision

/// A straight line.
struct Line {
  let from: CGPoint
  let to: CGPoint
}

/// Line connection
struct LineConnection {
  let color: UIColor
  let lines: [Line]
}

/**
 This structure holds the display parameters for the overlay to be drawon on a detected object.
 */
struct ObjectOverlay {
  let dots: [CGPoint]
  let lineConnectios: [LineConnection]
}

/// Custom view to visualize the face landmarks result on top of the input image.
class OverlayView: UIView {

  var objectOverlays: [ObjectOverlay] = []
  private let lineWidth: CGFloat = 3
  private let pointRadius: CGFloat = 3
  private let pointColor = UIColor.yellow

  override func draw(_ rect: CGRect) {
    for objectOverlay in objectOverlays {
      drawDots(objectOverlay.dots)
      for lineConnection in objectOverlay.lineConnectios {
        drawLines(lineConnection.lines, lineColor: lineConnection.color)
      }
    }
  }

  /**
   This method takes the landmarks, translates the points and lines to the current view, draws to the  of inferences.
   */
  func drawLandmarks(_ landmarks: [[NormalizedLandmark]], orientation: UIImage.Orientation, withImageSize imageSize: CGSize) {
    guard !landmarks.isEmpty else {
      objectOverlays = []
      setNeedsDisplay()
      return
    }

    var viewWidth = bounds.size.width
    var viewHeight = bounds.size.height
    var originX: CGFloat = 0
    var originY: CGFloat = 0

    if viewWidth / viewHeight > imageSize.width / imageSize.height {
      viewHeight = imageSize.height / imageSize.width  * bounds.size.width
      originY = (bounds.size.height - viewHeight) / 2
    } else {
      viewWidth = imageSize.width / imageSize.height * bounds.size.height
      originX = (bounds.size.width - viewWidth) / 2
    }

    var objectOverlays: [ObjectOverlay] = []

    for landmark in landmarks {
      var transformedLandmark: [CGPoint]!

      switch orientation {
      case .left:
        transformedLandmark = landmark.map({CGPoint(x: CGFloat($0.y), y: 1 - CGFloat($0.x))})
      case .right:
        transformedLandmark = landmark.map({CGPoint(x: 1 - CGFloat($0.y), y: CGFloat($0.x))})
      default:
        transformedLandmark = landmark.map({CGPoint(x: CGFloat($0.x), y: CGFloat($0.y))})
      }

      let dots: [CGPoint] = transformedLandmark.map({CGPoint(x: CGFloat($0.x) * viewWidth + originX, y: CGFloat($0.y) * viewHeight + originY)})
      var lineConnections: [LineConnection] = []
      lineConnections.append(LineConnection(
        color: UIColor(red: 0, green: 127/255.0, blue: 139/255.0, alpha: 1),
        lines: FaceLandmarker.faceOvalConnections()
          .map({ connection in
          let start = transformedLandmark[Int(connection.start)]
          let end = transformedLandmark[Int(connection.end)]
            return Line(from: CGPoint(x: CGFloat(start.x) * viewWidth + originX, y: CGFloat(start.y) * viewHeight + originY),
                        to: CGPoint(x: CGFloat(end.x) * viewWidth + originX, y: CGFloat(end.y) * viewHeight + originY))
          })))
      lineConnections.append(LineConnection(
        color: UIColor(red: 18/255.0, green: 181/255.0, blue: 203/255.0, alpha: 1),
        lines: FaceLandmarker.rightEyebrowConnections()
          .map({ connection in
          let start = transformedLandmark[Int(connection.start)]
          let end = transformedLandmark[Int(connection.end)]
            return Line(from: CGPoint(x: CGFloat(start.x) * viewWidth + originX, y: CGFloat(start.y) * viewHeight + originY),
                        to: CGPoint(x: CGFloat(end.x) * viewWidth + originX, y: CGFloat(end.y) * viewHeight + originY))
        })))
      lineConnections.append(LineConnection(
        color: UIColor(red: 18/255.0, green: 181/255.0, blue: 203/255.0, alpha: 1),
        lines: FaceLandmarker.leftEyebrowConnections()
        .map({ connection in
        let start = transformedLandmark[Int(connection.start)]
        let end = transformedLandmark[Int(connection.end)]
          return Line(from: CGPoint(x: CGFloat(start.x) * viewWidth + originX, y: CGFloat(start.y) * viewHeight + originY),
                      to: CGPoint(x: CGFloat(end.x) * viewWidth + originX, y: CGFloat(end.y) * viewHeight + originY))
      })))
      lineConnections.append(LineConnection(
        color: UIColor(red: 279/255.0, green: 171/255.0, blue: 0, alpha: 1),
        lines: FaceLandmarker.rightEyeConnections()
        .map({ connection in
        let start = transformedLandmark[Int(connection.start)]
        let end = transformedLandmark[Int(connection.end)]
          return Line(from: CGPoint(x: CGFloat(start.x) * viewWidth + originX, y: CGFloat(start.y) * viewHeight + originY),
                      to: CGPoint(x: CGFloat(end.x) * viewWidth + originX, y: CGFloat(end.y) * viewHeight + originY))
      })))
      lineConnections.append(LineConnection(
        color: UIColor(red: 279/255.0, green: 171/255.0, blue: 0, alpha: 1),
        lines: FaceLandmarker.leftEyeConnections()
        .map({ connection in
        let start = transformedLandmark[Int(connection.start)]
        let end = transformedLandmark[Int(connection.end)]
          return Line(from: CGPoint(x: CGFloat(start.x) * viewWidth + originX, y: CGFloat(start.y) * viewHeight + originY),
                      to: CGPoint(x: CGFloat(end.x) * viewWidth + originX, y: CGFloat(end.y) * viewHeight + originY))
      })))
      lineConnections.append(LineConnection(
        color: UIColor(red: 176/255.0, green: 0, blue: 32/255.0, alpha: 1),
        lines: FaceLandmarker.lipsConnections()
        .map({ connection in
        let start = transformedLandmark[Int(connection.start)]
        let end = transformedLandmark[Int(connection.end)]
          return Line(from: CGPoint(x: CGFloat(start.x) * viewWidth + originX, y: CGFloat(start.y) * viewHeight + originY),
                      to: CGPoint(x: CGFloat(end.x) * viewWidth + originX, y: CGFloat(end.y) * viewHeight + originY))
      })))
      objectOverlays.append(ObjectOverlay(dots: dots, lineConnectios: lineConnections))
    }

    self.objectOverlays = objectOverlays
    setNeedsDisplay()
  }

  private func drawDots(_ dots: [CGPoint]) {
    for dot in dots {
      let dotRect = CGRect(
        x: CGFloat(dot.x) - pointRadius / 2, y: CGFloat(dot.y) - pointRadius / 2,
        width: pointRadius, height: pointRadius)
      let path = UIBezierPath(ovalIn: dotRect)
      pointColor.setFill()
      path.fill()
    }
  }

  private func drawLines(_ lines: [Line], lineColor: UIColor) {
    let path = UIBezierPath()
    for line in lines {
      path.move(to: line.from)
      path.addLine(to: line.to)
    }
    path.lineWidth = lineWidth
    lineColor.setStroke()
    path.stroke()
  }
}

File: examples/pose_landmarker/android/app/src/main/java/com/google/mediapipe/examples/poselandmarker/OverlayView.kt
/*
 * Copyright 2023 The TensorFlow Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.mediapipe.examples.poselandmarker

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import androidx.core.content.ContextCompat
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import kotlin.math.max
import kotlin.math.min

class OverlayView(context: Context?, attrs: AttributeSet?) :
    View(context, attrs) {

    // Render mode: RGB_OVERLAY shows raw camera, DESENSITIZED draws mosaic background + skeleton
    enum class RenderMode {
        RGB_OVERLAY,
        DESENSITIZED
    }

    private var results: PoseLandmarkerResult? = null
    private var pointPaint = Paint()
    private var linePaint = Paint()

    // Desensitized mode fields
    var renderMode: RenderMode = RenderMode.RGB_OVERLAY
        set(value) {
            field = value
            invalidate()
        }

    // Current frame bitmap used to render mosaic background in desensitized mode
    private var currentBitmap: Bitmap? = null
    // Downscaled bitmap; drawing it back upscaled produces the pixelated mosaic effect
    private var mosaicBitmap: Bitmap? = null
    private var mosaicSrcRect = Rect()
    private var mosaicDstRectF = RectF()

    // Mosaic strength: smaller = coarser blocks (more desensitized), larger = clearer
    private val mosaicScale = 0.08f

    private var scaleFactor: Float = 1f
    private var imageWidth: Int = 1
    private var imageHeight: Int = 1

    // Offset mapping image coords to view coords (FILL_START crops the overflow)
    private var offsetX: Float = 0f
    private var offsetY: Float = 0f

    init {
        initPaints()
    }

    fun clear() {
        results = null
        currentBitmap = null
        mosaicBitmap = null
        pointPaint.reset()
        linePaint.reset()
        invalidate()
        initPaints()
    }

    private fun initPaints() {
        linePaint.color =
            ContextCompat.getColor(context!!, R.color.mp_color_primary)
        linePaint.strokeWidth = LANDMARK_STROKE_WIDTH
        linePaint.style = Paint.Style.STROKE

        pointPaint.color = Color.YELLOW
        pointPaint.strokeWidth = LANDMARK_STROKE_WIDTH
        pointPaint.style = Paint.Style.FILL
    }

    override fun draw(canvas: Canvas) {
        super.draw(canvas)
        if (renderMode == RenderMode.DESENSITIZED) {
            drawDesensitizedBackground(canvas)
        }
        drawPoseLandmarks(canvas)
    }

    /**\n     * Desensitized background: downscale the raw frame then draw it back upscaled\n     * (pixelated mosaic), then overlay a semi-transparent dark scrim to fully obscure\n     * face/clothing details while keeping body silhouette for skeleton alignment.\n     */
    private fun drawDesensitizedBackground(canvas: Canvas) {
        val bmp = currentBitmap ?: run {
            // No frame yet: use a solid dark background
            canvas.drawColor(Color.parseColor("#1A1A1A"))
            return
        }

        // Create or update the downscaled mosaic bitmap
        val targetW = max(1, (bmp.width * mosaicScale).toInt())
        val targetH = max(1, (bmp.height * mosaicScale).toInt())
        if (mosaicBitmap == null ||
            mosaicBitmap!!.width != targetW ||
            mosaicBitmap!!.height != targetH
        ) {
            mosaicBitmap?.recycle()
            mosaicBitmap = Bitmap.createScaledBitmap(bmp, targetW, targetH, false)
        } else {
            // Reuse the existing bitmap and refill pixels (cheaper than createScaledBitmap)
            val canvas2 = Canvas(mosaicBitmap!!)
            canvas2.drawBitmap(bmp, null, Rect(0, 0, targetW, targetH), null)
        }

        mosaicSrcRect = Rect(0, 0, targetW, targetH)
        // Draw the mosaic upscaled by the same scaleFactor as the skeleton, top-left aligned\n        // (matches PreviewView FILL_START). Overflow is clipped by the canvas, so skeleton\n        // and background stay perfectly aligned.
        val scaledW = imageWidth * scaleFactor
        val scaledH = imageHeight * scaleFactor
        mosaicDstRectF = RectF(0f, 0f, scaledW, scaledH)
        val mosaicPaint = Paint().apply {
            isFilterBitmap = false  // Keep pixelated blocks (no smoothing)
            isAntiAlias = false
        }
        canvas.drawBitmap(mosaicBitmap!!, mosaicSrcRect, mosaicDstRectF, mosaicPaint)

        // Overlay a semi-transparent dark scrim to further obscure details and make the skeleton pop
        val scrimPaint = Paint().apply {
            color = Color.argb(140, 10, 10, 10)  // ~55% opacity dark gray
            style = Paint.Style.FILL
        }
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), scrimPaint)
    }

    private fun drawPoseLandmarks(canvas: Canvas) {
        results?.let { poseLandmarkerResult ->
            for (landmark in poseLandmarkerResult.landmarks()) {
                for (normalizedLandmark in landmark) {
                    canvas.drawPoint(
                        normalizedLandmark.x() * imageWidth * scaleFactor + offsetX,
                        normalizedLandmark.y() * imageHeight * scaleFactor + offsetY,
                        pointPaint
                    )
                }

                PoseLandmarker.POSE_LANDMARKS.forEach {
                    canvas.drawLine(
                        poseLandmarkerResult.landmarks().get(0).get(it!!.start()).x() * imageWidth * scaleFactor + offsetX,
                        poseLandmarkerResult.landmarks().get(0).get(it.start()).y() * imageHeight * scaleFactor + offsetY,
                        poseLandmarkerResult.landmarks().get(0).get(it.end()).x() * imageWidth * scaleFactor + offsetX,
                        poseLandmarkerResult.landmarks().get(0).get(it.end()).y() * imageHeight * scaleFactor + offsetY,
                        linePaint)
                }
            }
        }
    }

    fun setResults(
        poseLandmarkerResults: PoseLandmarkerResult,
        imageHeight: Int,
        imageWidth: Int,
        runningMode: RunningMode = RunningMode.IMAGE,
        bitmap: Bitmap? = null
    ) {
        results = poseLandmarkerResults

        this.imageHeight = imageHeight
        this.imageWidth = imageWidth

        // Keep the current frame bitmap for desensitized background rendering
        currentBitmap = bitmap

        scaleFactor = computeScaleFactor(runningMode, width, height, imageWidth, imageHeight)

        // FILL_START aligns the scaled image to the top-left and crops the overflow.
        // Both skeleton and mosaic background use the same alignment, so they overlap exactly.
        offsetX = 0f
        offsetY = 0f

        invalidate()
    }

    companion object {
        private const val LANDMARK_STROKE_WIDTH = 12F

        /**
         * Compute the scale factor that maps normalized image coordinates to view coordinates.
         *
         * - IMAGE / VIDEO: fit the image inside the view (min ratio, letterboxed).
         * - LIVE_STREAM: fill the view like PreviewView FILL_START (max ratio, overflow cropped).
         *
         * Extracted as a pure function so it can be unit-tested without instantiating the View.
         */
        fun computeScaleFactor(
            runningMode: RunningMode,
            viewWidth: Int,
            viewHeight: Int,
            imageWidth: Int,
            imageHeight: Int
        ): Float {
            val w = viewWidth * 1f / imageWidth
            val h = viewHeight * 1f / imageHeight
            return when (runningMode) {
                RunningMode.IMAGE,
                RunningMode.VIDEO -> min(w, h)
                RunningMode.LIVE_STREAM -> max(w, h)
            }
        }
    }
}